src/Microsoft.ApplicationInsights.AspNetCore/TelemetryInitializers/AzureAppServiceRoleNameFromHostNameHeaderInitializer.cs

```diff
@@ -0,0 +1,149 @@
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+    using Microsoft.AspNetCore.Http;
+
+    /// <summary>
+    /// A telemetry initializer that will gather Azure Web App Role Environment context information to
+    /// populate TelemetryContext.Cloud.RoleName
+    /// This uses the http header "WAS-DEFAULT-HOSTNAME" to update role name, if available.
+    /// Otherwise role name is populated from "WEBSITE_HOSTNAME" environment variable.
+    /// </summary>
+    /// <remarks>
+    /// The RoleName is expected to contain the host name + slot name, but will be same across all instances of
+    /// a single App Service.
+    /// Populating RoleName from HOSTNAME environment variable will cause RoleName to be incorrect when a slot swap occurs in AppService.
+    /// The most accurate way to determine the RoleName is to rely on the header WAS-DEFAULT-HOSTNAME, as its
+    /// populated from App service front end on every request. Slot swaps are instantly reflected in this header.
+    /// </remarks>
+    public class AzureAppServiceRoleNameFromHostNameHeaderInitializer : ITelemetryInitializer
+    {
+        private const string WebAppHostNameHeaderName = "WAS-DEFAULT-HOSTNAME";
+        private const string WebAppHostNameEnvironmentVariable = "WEBSITE_HOSTNAME";
+        private readonly IHttpContextAccessor httpContextAccessor;
+        private string roleName;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureAppServiceRoleNameFromHostNameHeaderInitializer" /> class.
+        /// </summary>
+        /// <param name="httpContextAccessor">Accessor to provide HttpContext if available.</param>
+        public AzureAppServiceRoleNameFromHostNameHeaderInitializer(IHttpContextAccessor httpContextAccessor)
+        {
+            this.httpContextAccessor = httpContextAccessor ?? throw new ArgumentNullException(nameof(httpContextAccessor));
+
+            try
+            {
+                var result = Environment.GetEnvironmentVariable(WebAppHostNameEnvironmentVariable);
+
+                if (!string.IsNullOrEmpty(result) && result.EndsWith(this.WebAppSuffix, StringComparison.OrdinalIgnoreCase))
+                {
+                    result = result.Substring(0, result.Length - this.WebAppSuffix.Length);
+                }
+
+                this.roleName = result;
+            }
+            catch (Exception ex)
+            {
+                AspNetCoreEventSource.Instance.LogAzureAppServiceRoleNameFromHostNameHeaderInitializerWarning(ex.ToInvariantString());
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets suffix of website name. This must be changed when running in non public Azure region.
+        /// Default value (Public Cloud):  ".azurewebsites.net"
+        /// For US Gov Cloud: ".azurewebsites.us"
+        /// For Azure Germany: ".azurewebsites.de".
+        /// </summary>
+        public string WebAppSuffix { get; set; } = ".azurewebsites.net";
```

Comment:
For .NET Framework I presume you could configure from ApplicationInsights.config, but how do you configure it for .NET Core?

Reply:
Following should work for Asp.net core:
services.AddSingleton<ITelemetryInitializer>(new MyTelemetryInitializer() { fieldName = "myfieldName" });
(Agree its not mentioned in the doc  https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core#adding-telemetryinitializers, but i can add there)

There exists no other **easy** alternate to add a telemetry initializer with fields.

Reply:
This is not a new problem - we already have telemetryinitializers which has public properties.

Reply:
https://github.com/MicrosoftDocs/azure-docs/issues/40819